### PR TITLE
Use Kokkos to launch kernels in CUDAWrappers::MatrixFree 

### PR DESCRIPTION
**Pull request by Rombur**
This PR uses Kokkos to launch the kernels in CUDAWrappers::MatrixFree. Since we need the threads in a warp to work together, we need to use `Kokkos::TeamPolicy` and `Kokkos::TeamThreadRange`. This also allows us to use scratch memory. The biggest issue with this PR is that the new code is not backward compatible. The main reason is that we took advantage of `threadIdx.*` and `blockIdx.*` to compute indices. This is not allowed anymore and we now need to pass the indices to the functions. So we have an extra argument to several functions.

### CODE REVIEW

File: include/deal.II/matrix_free/cuda_matrix_free.h [hunk 1]
```diff
@@ -628,114 +636,48 @@ namespace CUDAWrappers
...
+      Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+    using SharedView2D = Kokkos::View<
+      Number *[dim],
+      MemorySpace::Default::kokkos_space::execution_space::scratch_memory_space,
+      Kokkos::MemoryTraits<Kokkos::Unmanaged>>;
+
+    DEAL_II_HOST_DEVICE
+    SharedData(const TeamHandle &team_member,
+               SharedView1D      values,
+               SharedView2D      gradients)
```
Comment (masterleinad) [inline include/deal.II/matrix_free/cuda_matrix_free.h:657]:
```suggestion
               const SharedView1D     &values,
               const SharedView2D     &gradients)
```
for consistency.

**Comment by masterleinad** [comment]
I think it's OK that the changes are not backward-compatible given that it's not too difficult for users to adapt and the `Kokkos`-based version allows for more flexibility.

Have you compared performance with the version before all the `Kokkos`-related changes?

**Comment by Rombur** [comment]
> Have you compared performance with the version before all the Kokkos-related changes?

Not yet. I have a following PR that fixes the shortcut I used when moving the data structures to Kokkos. Right now during setup, we are copying the data to the gpu one cell at the time.

**Review by masterleinad — APPROVED** [review]
Looms OK to me.

**Pull request merged**